openwisp_monitoring/device/apps.py

```diff
@@ -11,3 +15,57 @@ class DeviceMonitoringConfig(AppConfig):
 
     def ready(self):
         manage_short_retention_policy()
+        from openwisp_controller.connection.models import DeviceConnection
+        from openwisp_controller.connection.signals import is_working_changed
+        is_working_changed.connect(self.health_status_changed,
+                                   sender=DeviceConnection,
+                                   dispatch_uid='change_health_status')
+
+    @classmethod
+    def health_status_changed(cls, instance, is_working, **kwargs):
+        from ..check.models import Check
+        d = instance.device
+        dm = d.monitoring
+        initial_status = dm.status
+        status = 'ok' if is_working else 'problem'
+        if not is_working:
+            dm.update_status(status)
+            # Create a related notification explaining why it's not working
+            desc = instance.failure_reason
+            opts = dict(actor=d, level='warning', verb='not working', description=desc)
+            if initial_status == 'ok':
+                try:
+                    # if checks exist trigger them
+                    Check.objects.get(object_id=d.pk)
+                    management.call_command('run_checks')
+                except Check.DoesNotExist:
+                    pass
+        else:
+            # create a notification that device is working
+            opts = dict(actor=d, level='info', verb='connected successfully')
+            try:
+                # if checks exist trigger them else, set status as 'ok'
+                Check.objects.get(object_id=instance.device.pk)
+                management.call_command('run_checks')
+            except Check.DoesNotExist:
+                dm.update_status(status)
+        opts['data'] = {'email_subject': f'[{status.upper()}] {d}'}
+        cls._notify_users(opts)
+
+    def _notify_users(opts):
+        User = get_user_model()
+        where = Q(is_superuser=True)
+        # only retrieve users which have the receive flag active
+        where = where & Q(notificationuser__receive=True)
+        # perform query
+        qs = (
+            User.objects.select_related('notificationuser')
+            .order_by('date_joined')
+            .filter(where)
+        )
+        Notification = load_model('notifications', 'Notification')
+        for user in qs:
+            n = Notification(**opts)
+            n.recipient = user
+            n.full_clean()
+            n.save()
```

Comment:
Did you considered using `notify` signal as mentioned in the [django-notifications documentation](https://github.com/django-notifications/django-notifications#generating-notifications)?

Reply:
Yes, spent some good time understanding and deciding upon it. The reason I have done it in this manner is as I wanted to maintain consistency with what has been done already. Anyhow, this reminds me that it would be best if we can have such a function in the notifications app maybe notification module is going to have an independent existence for best :smile: